File: packages/node-peer/src/http-sender-transport.ts
/**
 * Node {@link SenderTransport} using the global `fetch` (backed by undici,
 * which pools connections per origin), so parallel chunk POSTs reuse sockets.
 */

import { API_NAMESPACE } from '@wrapdrive/protocol';
import type {
  ChunkUpload,
  SenderTransport,
  TransportResponse,
  UploadTarget,
} from '@wrapdrive/transfer-engine';

/** Posts chunks/streams to a peer's WrapDrive endpoints over HTTP. */
export class HttpSenderTransport implements SenderTransport {
  constructor(private readonly baseUrl: string) {}

  async uploadChunk(upload: ChunkUpload): Promise<TransportResponse> {
    const url =
      `${this.baseUrl}${API_NAMESPACE}/upload-chunk` +
      `?sessionId=${encodeURIComponent(upload.sessionId)}` +
      `&fileId=${encodeURIComponent(upload.fileId)}` +
      `&token=${encodeURIComponent(upload.token)}` +
      `&chunkIndex=${upload.chunkIndex}` +
      `&offset=${upload.offset}` +
      `&length=${upload.length}`;
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/octet-stream' },
      body: upload.data,
    });
    return { status: res.status };
  }

  async uploadStream(
    target: UploadTarget,
    body: AsyncIterable<Uint8Array>,
  ): Promise<TransportResponse> {
    const url =
      `${this.baseUrl}${API_NAMESPACE}/upload` +
      `?sessionId=${encodeURIComponent(target.sessionId)}` +
      `&fileId=${encodeURIComponent(target.fileId)}` +
      `&token=${encodeURIComponent(target.token)}`;
    const chunks: Uint8Array[] = [];
    for await (const c of body) chunks.push(c);
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/octet-stream' },
      body: Buffer.concat(chunks),
    });
    return { status: res.status };
  }

  async cancel(sessionId: string): Promise<void> {
    await fetch(`${this.baseUrl}${API_NAMESPACE}/cancel?sessionId=${sessionId}`, {
      method: 'POST',
    }).catch(() => undefined);
  }
}

File: apps/android/app/src/main/kotlin/com/wrapdrive/net/HttpSenderTransport.kt
package com.wrapdrive.net

import com.wrapdrive.core.protocol.WrapDriveProtocol
import com.wrapdrive.core.transfer.ChunkUpload
import com.wrapdrive.core.transfer.SenderTransport
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.contentType

/**
 * [SenderTransport] backed by the Ktor CIO client. Posts chunks to the peer's
 * `/upload-chunk` endpoint and supports `/cancel`. Connection pooling/keep-alive
 * is handled by the shared client.
 */
class HttpSenderTransport(
    private val baseUrl: String, // e.g. http://192.168.1.5:53317
) : SenderTransport {
    private val client = HttpClient(CIO)
    private val ns = WrapDriveProtocol.API_NAMESPACE

    override suspend fun uploadChunk(upload: ChunkUpload): Int {
        val url =
            "$baseUrl$ns/upload-chunk" +
                "?sessionId=${upload.target.sessionId}" +
                "&fileId=${upload.target.fileId}" +
                "&token=${upload.target.token}" +
                "&chunkIndex=${upload.chunkIndex}" +
                "&offset=${upload.offset}" +
                "&length=${upload.length}"
        val response: HttpResponse =
            client.post(url) {
                contentType(ContentType.Application.OctetStream)
                setBody(upload.data)
            }
        return response.status.value
    }

    override suspend fun cancel(sessionId: String) {
        runCatching { client.post("$baseUrl$ns/cancel?sessionId=$sessionId") }
    }

    fun close() {
        client.close()
    }
}
